File: 14_longest_common_prefix/main.cpp
/*
 * Write a function to find the longest common prefix string amongst an array of strings.
 * If there is no common prefix, return an empty string "".

 * Example 1:
 * Input: ["flower","flow","flight"]
 * Output: "fl"
 *
 * Example 2:
 * Input: ["dog","racecar","car"]
 * Output: ""
 * Explanation: There is no common prefix among the input strings.
 *
 * Note: All given inputs are in lowercase letters a-z.
 */

#include <iostream>
#include <vector>

class Solution {
public:
    std::string longestCommonPrefix(std::vector<std::string>& strs) {
	if (strs.size() == 0) return "";

	for (int i = 0; i < strs[0].size(); ++i) {
	    for (int j = 1; j < strs.size(); ++j) {
		if (i >= strs[j].size() || strs[j][i] != strs[0][i]) {
		    return strs[0].substr(0, i);
		}
	    }
	}
	return strs[0];
    }
};

int main() {
    Solution s;

    std::vector<std::string> strs = {"flower", "flow", "flight"};
    std::cout << s.longestCommonPrefix(strs) << std::endl;

    strs = {"dog", "racecar", "car"};
    std::cout << s.longestCommonPrefix(strs) << std::endl;

    strs = {};
    std::cout << s.longestCommonPrefix(strs) << std::endl;
}

File: 14_longest_common_prefix/main.py
# -*- coding: utf-8 -*-


class Solution(object):

    def longest_common_prefix(self, strs):
        if len(strs) == 0: return ''
        for i in range(0, len(strs[0])):
            for j in range(1, len(strs)):
                if i >= len(strs[j]) or strs[j][i] != strs[0][i]:
                    return strs[0][0:i]
        return strs[0]


if __name__ == '__main__':
    s = Solution()

    strs = ['flower', 'flow', 'flight']
    print(s.longest_common_prefix(strs))

    strs = ['dog', 'racecar', 'car']
    print(s.longest_common_prefix(strs))

    strs = ['aa', 'a']
    print(s.longest_common_prefix(strs))
